sys/kern/subr_bus.c

```diff
@@ -5678,13 +5713,54 @@ static int
 devctl2_ioctl(struct cdev *cdev, u_long cmd, caddr_t data, int fflag,
     struct thread *td)
 {
+#if defined(COMPAT_FREEBSD32) || defined(COMPAT_FREEBSD64)
+	struct devreq dr;
+#endif
+#ifdef COMPAT_FREEBSD32
+	struct devreq32 *req32;
+#endif
+#ifdef COMPAT_FREEBSD64
+	struct devreq64 *req64;
+#endif
 	struct devreq *req;
 	device_t dev;
 	int error, old;
 
+#ifdef COMPAT_FREEBSD32
+	/*
+	 * NOTE: the compat code relies on all handled commands being
+	 * _IOW() defines with the size of struct devreq.  If other
+	 * structs are added or _IOR is used, this code needs to
+	 * be adjusted accordingly.
+	 *
+	 * No consumers of dr_buffer exist so we always copy dr_data
+	 * and leave dr_buffer.length zeroed.
+	 */
+	if (IOCPARM_LEN(cmd) == sizeof(struct devreq32)) {
+		req32 = (struct devreq32 *)data;
+		memset(&dr, 0, sizeof(dr));
+		memcpy(dr.dr_name, req32->dr_name, sizeof(dr.dr_name));
+		CP(*req32, dr, dr_flags);
+		PTRIN_CP(*req32, dr, dr_data);
+		req = &dr;
+		cmd = _IOC_NEWTYPE(cmd, struct devreq);
+	} else
+#endif
+#ifdef COMPAT_FREEBSD64
+	if (IOCPARM_LEN(cmd) == sizeof(struct devreq64)) {
+		req64 = (struct devreq64 *)data;
+		memset(&dr, 0, sizeof(dr));
+		memcpy(dr.dr_name, req64->dr_name, sizeof(dr.dr_name));
+		CP(*req64, dr, dr_flags);
+		dr.dr_data = __USER_CAP_UNBOUND(req64->dr_data);
```

Comment:
This could use __USER_CAP_STRING() or whatever we have?  I think it is still unbound in practice, but perhaps documents it better?

Reply:
Switched. I've had a mental todo item forever to replace move instances of __USER_CAP_STR with a __USER_CAP_PATH that can use MAXPATHLEN as their bounds.